File: apps/mobile/compose/src/commonTest/kotlin/co/typie/screen/editor/editor/TrackedRangeMembershipTest.kt
package co.typie.screen.editor.editor

import co.typie.editor.ffi.Affinity
import co.typie.editor.ffi.Position
import co.typie.editor.ffi.TrackedRangeEndpoints
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class TrackedRangeMembershipTest {
  private val position = Position(node = "paragraph", offset = 1, affinity = Affinity.Downstream)

  private fun range(id: String, group: String) =
    TrackedRangeEndpoints(id = id, group = group, anchor = position, head = position)

  @Test
  fun eligibleActiveMemberWinsWithoutChangingCoreFallbackOrder() {
    val members = listOf(range("before", "comment"), range("after", "comment-active"))

    assertEquals(
      "after",
      members
        .selectTrackedRangeMember(
          allowedGroups = setOf("comment", "comment-active"),
          activeId = "after",
        )
        ?.id,
    )
    assertEquals(
      "before",
      members
        .selectTrackedRangeMember(
          allowedGroups = setOf("comment", "comment-active"),
          activeId = null,
        )
        ?.id,
    )
  }

  @Test
  fun excludedOrUnownedActiveIdFallsBackToFirstEligibleMember() {
    val members =
      listOf(
        range("temporary", "__comment_compose__"),
        range("missing-result", "spellcheck-active"),
        range("owned-result", "spellcheck"),
      )

    assertEquals(
      "owned-result",
      members
        .selectTrackedRangeMember(
          allowedGroups = setOf("spellcheck", "spellcheck-active"),
          activeId = "missing-result",
          ownedIds = setOf("owned-result"),
        )
        ?.id,
    )
    assertNull(
      members.selectTrackedRangeMember(
        allowedGroups = setOf("comment", "comment-active"),
        activeId = "temporary",
      )
    )
    assertNull(
      listOf(range("orphan", "comment"))
        .selectTrackedRangeMember(
          allowedGroups = setOf("comment", "comment-active"),
          activeId = "orphan",
          ownedIds = emptySet(),
        )
    )
  }

  @Test
  fun membershipIdsIgnoreNormalActiveGroupProjection() {
    val groups = setOf("comment", "comment-active")

    val before =
      listOf(range("a", "comment"), range("b", "comment-active")).trackedRangeMembershipIds(groups)
    val after =
      listOf(range("a", "comment-active"), range("b", "comment")).trackedRangeMembershipIds(groups)

    assertEquals(before, after)
  }
}

File: apps/website/src/lib/editor-ffi/tracked-range-membership.test.ts
import { describe, expect, it, vi } from 'vitest';
import { selectTrackedRangeMember, semanticMembershipForStateChange, trackedRangeMembershipIds } from './tracked-range-membership';
import type { Position, Selection, StateField, TrackedRangeEndpoints } from '@typie/editor-ffi/browser';

const position = (offset: number): Position => ({ node: 'paragraph', offset, affinity: 'downstream' });
const collapsed = (offset: number): Selection => ({ anchor: position(offset), head: position(offset) });
const member = (id: string, group: string): TrackedRangeEndpoints => ({
  id,
  group,
  anchor: position(0),
  head: position(1),
});

describe('tracked range membership', () => {
  it.each<StateField>(['selection', 'doc', 'tracked_ranges'])('%s change refreshes membership under a stationary cursor', (field) => {
    const query = vi.fn(() => [member('a', 'comment')]);

    const selection = collapsed(3);
    expect(semanticMembershipForStateChange(new Set([field]), selection, query)).toEqual([member('a', 'comment')]);
    expect(query).toHaveBeenCalledWith(selection);
  });

  it('queries membership for non-collapsed selections', () => {
    const query = vi.fn(() => [member('a', 'comment')]);
    const expanded: Selection = { anchor: position(1), head: position(2) };

    expect(semanticMembershipForStateChange(new Set<StateField>(['selection']), expanded, query)).toEqual([member('a', 'comment')]);
    expect(query).toHaveBeenCalledWith(expanded);
  });

  it('does not query when the selection is unavailable', () => {
    const query = vi.fn(() => [member('a', 'comment')]);

    expect(semanticMembershipForStateChange(new Set<StateField>(['selection']), undefined, query)).toBeUndefined();
    expect(query).not.toHaveBeenCalled();
  });

  it('does not query for layout-only changes', () => {
    const query = vi.fn(() => [member('a', 'comment')]);

    expect(semanticMembershipForStateChange(new Set<StateField>(['page_sizes', 'cursor']), collapsed(1), query)).toBeUndefined();
    expect(query).not.toHaveBeenCalled();
  });

  it('prefers an eligible active ID, otherwise keeps Core fallback order', () => {
    const members = [member('before', 'comment'), member('after', 'comment-active')];
    const groups = new Set(['comment', 'comment-active']);

    expect(selectTrackedRangeMember(members, groups, 'after')?.id).toBe('after');
    expect(selectTrackedRangeMember(members, groups, null)?.id).toBe('before');
    expect(selectTrackedRangeMember(members, groups, 'outside')?.id).toBe('before');
  });

  it('filters feature ownership before applying active preference', () => {
    const members = [member('missing', 'spellcheck-active'), member('owned', 'spellcheck')];

    expect(selectTrackedRangeMember(members, new Set(['spellcheck', 'spellcheck-active']), 'missing', new Set(['owned']))?.id).toBe(
      'owned',
    );
  });

  it('keeps membership identity stable across normal-active group projection', () => {
    const groups = new Set(['comment', 'comment-active']);

    expect(trackedRangeMembershipIds([member('a', 'comment'), member('b', 'comment-active')], groups)).toEqual(
      trackedRangeMembershipIds([member('a', 'comment-active'), member('b', 'comment')], groups),
    );
  });
});
